servant-server/src/Servant/Server/Internal/Router.hs

```diff
@@ -1,25 +1,40 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DeriveFunctor #-}
+
 module Servant.Server.Internal.Router where
 
+#if !MIN_VERSION_base(4,8,0)
+import           Control.Applicative         ((<$>))
+#endif
 import           Data.Map                                   (Map)
 import qualified Data.Map                                   as M
 import           Data.Monoid                                ((<>))
 import           Data.Text                                  (Text)
-import           Network.Wai                                (Request, pathInfo)
+import           Network.Wai                                (Request, Response, pathInfo)
 import           Servant.Server.Internal.PathInfo
 import           Servant.Server.Internal.RoutingApplication
 
+type Router = Router' RoutingApplication
+
 -- | Internal representation of a router.
-data Router =
+data Router' a =
     WithRequest   (Request -> Router)
       -- ^ current request is passed to the router
   | StaticRouter  (Map Text Router)
       -- ^ first path component used for lookup and removed afterwards
   | DynamicRouter (Text -> Router)
       -- ^ first path component used for lookup and removed afterwards
-  | LeafRouter    RoutingApplication
+  | LeafRouter    a
       -- ^ to be used for routes that match an empty path
   | Choice        Router Router
       -- ^ left-biased choice between two routers
+  deriving Functor
+
+-- | Apply a function to the result of a router in functor style.  The result contains the failure
```

Comment:
This haddock isn't very clear to me. "Functor style" doesn't seem helpful, and it's not clear to me why *middleware response values* means, or why only failures are mentioned. Why not just "Apply the transformation to the response of a `Router`"?